frame/support/src/storage/types/key.rs

```diff
@@ -0,0 +1,900 @@
+/// Hasher1(encode(key1))
+///     ++ Hasher2(encode(key2))
+///     ++ ...
+///     ++ HasherN(encode(keyN))
+/// ```
+pub struct Key<Hasher, KeyType>(core::marker::PhantomData<(Hasher, KeyType)>);
+
+/// A trait that contains the current key as an associated type.
+pub trait KeyGenerator {
+    type Key: EncodeLike<Self::Key>;
+    type KArg: Encode;
+    type HashFn: FnOnce(&[u8]) -> Vec<u8>;
+    type HArg;
+
+    fn final_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(key: KArg) -> Vec<u8>;
+    fn migrate_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(
+        key: &KArg,
+        hash_fns: Self::HArg,
+    ) -> Vec<u8>;
+    fn final_hash(encoded: &[u8]) -> Vec<u8>;
+}
+
+impl<H: StorageHasher, K: FullCodec> KeyGenerator for Key<H, K> {
+    type Key = K;
+    type KArg = (K,);
+    type HashFn = Box<dyn FnOnce(&[u8]) -> Vec<u8>>;
+    type HArg = (Self::HashFn,);
+
+    fn final_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(key: KArg) -> Vec<u8> {
+        H::hash(
+            &key.to_encoded_iter()
+                .next()
+                .expect("should have at least one element!"),
+        )
+        .as_ref()
+        .to_vec()
+    }
+
+    fn migrate_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(
+        key: &KArg,
+        hash_fns: Self::HArg,
+    ) -> Vec<u8> {
+        (hash_fns.0)(
+            &key.to_encoded_iter()
+                .next()
+                .expect("should have at least one element!"),
+        )
+    }
+
+    fn final_hash(encoded: &[u8]) -> Vec<u8> {
+        H::hash(encoded).as_ref().to_vec()
+    }
+}
+
+#[impl_trait_for_tuples::impl_for_tuples(2, 18)]
+impl KeyGenerator for Tuple {
+    for_tuples!( type Key = ( #(Tuple::Key),* ); );
+    for_tuples!( type KArg = ( #(Tuple::Key),* ); );
+    for_tuples!( type HArg = ( #(Tuple::HashFn),* ); );
+    type HashFn = Box<dyn FnOnce(&[u8]) -> Vec<u8>>;
+
+    fn final_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(key: KArg) -> Vec<u8> {
+        let mut final_key = Vec::new();
+        let mut iter = key.to_encoded_iter();
+        for_tuples!(
+            #(
+                let next_encoded = iter.next().expect("KArg number should be equal to Key number");
+                final_key.extend_from_slice(&Tuple::final_hash(&next_encoded));
+            )*
+        );
+        final_key
+    }
+
+    fn migrate_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(
+        key: &KArg,
+        hash_fns: Self::HArg,
+    ) -> Vec<u8> {
+        let mut migrated_key = Vec::new();
+        let mut iter = key.to_encoded_iter();
+        for_tuples!(
+            #(
+                let next_encoded = iter.next().expect("KArg number should be equal to Key number");
+                migrated_key.extend_from_slice(&(hash_fns.Tuple)(&next_encoded));
+            )*
+        );
+        migrated_key
+    }
+
+    fn final_hash(encoded: &[u8]) -> Vec<u8> {
+        panic!("final_hash called on KeyGenerator tuple")
```

Comment:
I think we can enforce this check at compile time.

What we want is that the element of the tuple provide final_hash. So at the end what we want is to bound more on the element that on the tuple itself.

In the current shape the storage:
```rust
		type CASASSA = StorageNMap<
			Prefix,
			(
				Key<Blake2_128Concat, u16>,
				(Key<Blake2_128Concat, u16>,
				Key<Twox64Concat, u16>),
			),
			u32,
			OptionQuery,
		>;
```
Is compiling fine, but panic at execution.

Instead I think we can improve by doing this:
```diff
diff --git a/frame/support/src/storage/types/key.rs b/frame/support/src/storage/types/key.rs
index f1cd07d302..e1d7c21d84 100755
--- a/frame/support/src/storage/types/key.rs
+++ b/frame/support/src/storage/types/key.rs
@@ -45,6 +45,10 @@ pub trait KeyGenerator {
         key: &KArg,
         hash_fns: Self::HArg,
     ) -> Vec<u8>;
+}
+
+/// A trait that contains the current key as an associated type.
+pub trait KeyGeneratorInner: KeyGenerator {
     fn final_hash(encoded: &[u8]) -> Vec<u8>;
 }
 
@@ -74,13 +78,17 @@ impl<H: StorageHasher, K: FullCodec> KeyGenerator for Key<H, K> {
                 .expect("should have at least one element!"),
         )
     }
+}
 
+
+impl<H: StorageHasher, K: FullCodec> KeyGeneratorInner for Key<H, K> {
     fn final_hash(encoded: &[u8]) -> Vec<u8> {
         H::hash(encoded).as_ref().to_vec()
     }
 }
 
 #[impl_trait_for_tuples::impl_for_tuples(2, 18)]
+#[tuple_types_custom_trait_bound(KeyGeneratorInner)]
 impl KeyGenerator for Tuple {
     for_tuples!( type Key = ( #(Tuple::Key),* ); );
     for_tuples!( type KArg = ( #(Tuple::Key),* ); );
@@ -113,10 +121,6 @@ impl KeyGenerator for Tuple {
         );
         migrated_key
     }
-
-    fn final_hash(encoded: &[u8]) -> Vec<u8> {
-        panic!("final_hash called on KeyGenerator tuple")
-    }
 }
 
 /// Trait to indicate that a tuple can be converted into an iterator of a vector of encoded bytes.
@@ -158,6 +162,7 @@ impl<H: ReversibleStorageHasher, K: FullCodec> ReversibleKeyGenerator for Key<H,
 }
 
 #[impl_trait_for_tuples::impl_for_tuples(2, 18)]
+#[tuple_types_custom_trait_bound(ReversibleKeyGenerator + KeyGeneratorInner)]
 impl ReversibleKeyGenerator for Tuple {
     for_tuples!( type Hasher = ( #(Tuple::Hasher),* ); );
 
```

Apart from this I think traits are in very good shape.